### DISCUSSION TITLE
Custom Route Loader: tagging with 'routing.route_loader' does not work!

### DISCUSSION THREAD

**Discussion by rafaelschmitt** [symfony/symfony#46025] [Q&A] [answered]
### Symfony version(s) affected

5.4

### Description

We tried to implement our own Custom Route Loader. We stuck to this documentary:
https://symfony.com/doc/5.4/routing/custom_route_loader.html

**However, on the one hand it says:**

> If you're using [autoconfigure](https://symfony.com/doc/5.4/service_container.html#services-autoconfigure), your class should implement the [RouteLoaderInterface](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Bundle/FrameworkBundle/Routing/RouteLoaderInterface.php) interface to be tagged automatically. If you're not using autoconfigure, tag it manually with routing.route_loader.

We also saw that the `RouteLoaderInterface` is linked to the tag `routing.route_loader`:
https://github.com/symfony/framework-bundle/blob/5.4/DependencyInjection/FrameworkExtension.php#L633

Additionally, we also found this:
https://github.com/symfony/framework-bundle/blob/5.4/CHANGELOG.md?plain=1#L98

**On the other hand, the example says this:**

> Notice the tag routing.loader. All services with this tag will be marked as potential route loaders and added as specialized route loaders to the routing.loader service, which is an instance of [DelegatingLoader](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php)

**So we have these 3 options:**
1. Implement `RouteLoaderInterface` in our Custom Route Loader class (which is linked with the tag `routing.route_loader`), when using `autoconfigure: true`
2. define the service manually and tag it with `routing.route_loader`
3. define the service manually and tag it with `routing.loader`

**=> 1 and 2 does not work! Only option 3 works. Why?**
 

### How to reproduce

First, create a symfony application:
https://symfony.com/doc/5.4/setup.html#creating-symfony-applications

Then, create/edit the following files:

`config/services.yaml`
```yaml
#...
services:
    _defaults:
        autowire: true
        autoconfigure: true
#...
```

`src/Routing/Loader/LocalizedRouteLoader.php`
```php
<?php

namespace App\Routing\Loader;

use Symfony\Bundle\FrameworkBundle\Routing\RouteLoaderInterface;
use Symfony\Component\Config\Loader\Loader;
use Symfony\Component\Routing\RouteCollection;

class LocalizedRouteLoader extends Loader implements RouteLoaderInterface
{
    public function load($resource, $type = null)
    {
        var_dump("I should be called!!!");
        return new RouteCollection();
    }

    public function supports($resource, string $type = null)
    {
        return 'localized' === $type;
    }
}
```

`src/Routing/routes.yaml`
```yaml
# empty file
```

`config/routes.yaml`
```yaml
app_routing:
    resource: '../src/Routing/routes.yaml'
    type: localized
```

**Option 1:**

Because `LocalizedRouteLoader` implements `RouteLoaderInterface` and `autoconfigure: true`, all should be fine.

Execute: `bin/console cache:clear`
Output: failed!
```
#...
Cannot load resource "../src/Routing/routes.yaml". Make sure there is a loader supporting the "localized" type.
#...
```

**Option 2:**

Define the service manually and tag it with `routing.route_loader`.

`config/services.yaml`
```yaml
#...
    App\Routing\Loader\LocalizedRouteLoader:
        tags: [routing.route_loader]
#...
```
Execute: `bin/console cache:clear`
Output: failed!
```
#...
Cannot load resource "../src/Routing/routes.yaml". Make sure there is a loader supporting the "localized" type.
#...
```

**Option 3:**

Define the service manually and tag it with `routing.loader`.

`config/services.yaml`
```yaml
#...
    App\Routing\Loader\LocalizedRouteLoader:
        tags: [routing.loader]
#...
```
Execute: `bin/console cache:clear`
Output: It works! The `var_dump` is called!
```
#...
/home/rafaelschmitt/projects/php/test/src/Routing/Loader/LocalizedRouteLoader.php:13:
string(21) "I should be called!!!"
#...
```

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce your issue?

**Comment by fancyweb** [comment, marked as answer]
Route loader and service route loader are two different things. In your case, `LocalizedRouteLoader` is actually a route loader and needs the `routing.loader` tag. There is no marker interface for this tag so autoconfiguration won't work.

The `RouteLoaderInterface` marker interface and the `routing.route_loader` tag are for service route loaders (`type: "service"` in the routes conf).

**Comment by rafaelschmitt** [comment]
@fancyweb Ah ok, I understand. Thank you for the clarification! So of course it's not a bug!
